main/src/ca/uwaterloo/flix/language/phase/CodegenHelper.scala

```diff
@@ -100,49 +100,107 @@ object CodegenHelper {
+    * All these classes include a field representing `tag`
+    */
+  sealed trait EnumCaseName extends QualName {
+    def tag : String
+  }
+
+  /**
+    * Qualified name of an enum case class
     * @param sym symbol of the enum
     * @param tag tag of the enum
     * @param tpe tpe of the field of the enum
     */
-  case class EnumClassName(sym: EnumSym, tag: String, tpe: WrappedType) extends QualName {
+  case class SECClassName(sym: EnumSym, tag: String, tpe: WrappedType) extends EnumCaseName {
     val ref: List[String] = tpe match {
-      case WrappedPrimitive(t) => CodegenHelper.fixedEnumPrefix ::: sym.namespace ::: List(sym.name, typeSpecifier(t), tag)
-      case _ => CodegenHelper.fixedEnumPrefix ::: sym.namespace ::: List(sym.name, "object", tag)
+      case WrappedPrimitive(t) => CodegenHelper.fixedEnumPrefix ::: sym.namespace ::: List(sym.name, "clazz", typeSpecifier(t), tag)
+      case _ => CodegenHelper.fixedEnumPrefix ::: sym.namespace ::: List(sym.name, "clazz", "obj", tag)
+    }
+  }
+
+  /**
+    * Qualified name of an enum case class
+    * @param sym symbol of the enum
+    * @param tag tag of the enum
+    * @param fields tpe of the field of the enum
+    */
+  case class ETFClassName(sym: EnumSym, tag: String, fields: List[WrappedType]) extends EnumCaseName {
+    private val desc: List[String] = fields.map{
+      case WrappedPrimitive(t) => typeSpecifier(t)
+      case _ => "obj"
     }
+
+    val ref: List[String] = CodegenHelper.fixedEnumPrefix ::: sym.namespace ::: List(sym.name, "fusionClazz") ::: desc ::: List(tag)
   }
 
   /**
     * Qualified name of an enum interface
     *
     * @param sym symbol of the enum
     */
-  case class EnumInterfName(sym: EnumSym) extends QualName {
-    val ref: List[String] = CodegenHelper.fixedEnumPrefix ::: sym.namespace ::: List(sym.name, "EnumInterface")
+  case class EnumTypeInterfaceName(sym: EnumSym) extends QualName {
+    val ref: List[String] = CodegenHelper.fixedEnumPrefix ::: sym.namespace ::: List(sym.name, "EnumTypeInterface")
   }
 
   /**
     * Qualified name of a tuple class
     *
     * @param fields fields of the tuple
     */
+  case class TupleInterfaceName(fields: List[WrappedType]) extends QualName {
+    val ref: List[String] = fixedTuplePrefix ::: fields.map{
+      case WrappedPrimitive(tpe) => typeSpecifier(tpe)
+      case WrappedNonPrimitives(_) => "object"
+    } ::: List("interf", "Tuple")
+  }
+
+  /**
+    * Qualified name of a tuple class
+    * @param fields fields of the tuple
+    */
   case class TupleClassName(fields: List[WrappedType]) extends QualName {
     val ref: List[String] = fixedTuplePrefix ::: fields.map {
       case WrappedPrimitive(tpe) => typeSpecifier(tpe)
       case WrappedNonPrimitives(_) => "object"
-    } ::: List("Tuple")
+    } ::: List("clazz", "Tuple")
   }
 
-  /**
+  def exceptionThrowerMethod(visitor: ClassWriter,
```

Comment:
This is a method with a reasonable number of arguments that are understandable.

Reply:
That sound sarcastic in many ways :D Anyway, I added a comment there just to keep the format.